src/blueapi/service/authentication.py

```diff
@@ -0,0 +1,180 @@
+    def load_token(token) -> dict[str, Any]: ...
+    @abstractmethod
+    def delete_token(self) -> None: ...
+
+
+class CliTokenManager(TokenManager):
+    def __init__(self, token_path: Path) -> None:
+        self._token_path: Path = token_path
+
+    @cached_property
+    def _file_path(self) -> str:
+        return os.path.expanduser(self._token_path)
+
+    def save_token(self, token: dict[str, Any]) -> None:
+        token_json: str = json.dumps(token)
+        token_base64: bytes = base64.b64encode(token_json.encode("utf-8"))
+        with open(self._file_path, "wb") as token_file:
+            token_file.write(token_base64)
+
+    def load_token(self) -> dict[str, Any]:
+        if not os.path.exists(self._file_path):
+            raise FileNotFoundError
+        with open(self._file_path, "rb") as token_file:
+            token_base64: bytes = token_file.read()
+            token_json: str = base64.b64decode(token_base64).decode("utf-8")
+            return json.loads(token_json)
+
+    def delete_token(self) -> None:
+        Path(self._file_path).unlink(missing_ok=True)
+
+
+class SessionManager:
+    def __init__(
+        self,
+        server_config: OIDCConfig,
+    ) -> None:
+        self._server_config = server_config
+        assert isinstance(
+            server_config, CLIClientConfig
+        ), "Please provide token_path in config"
+        self._token_manager: TokenManager = CliTokenManager(server_config.token_path)
+
+    @cached_property
+    def client(self):
+        return jwt.PyJWKClient(self._server_config.jwks_uri)
+
+    def get_token(self) -> dict[str, Any]:
+        return self._token_manager.load_token()
+
+    def logout(self) -> None:
+        try:
+            token = self._token_manager.load_token()
+            requests.get(
+                self._server_config.end_session_endpoint,
+                params={
+                    "id_token_hint": token["id_token"],
+                    "client_id": self._server_config.client_id,
+                },
+            )
+        except Exception:
+            ...  # Swallow any exceptions
+        finally:
+            self._token_manager.delete_token()
+
+    def decode_jwt(self, json_web_token: str):
+        signing_key = self.client.get_signing_key_from_jwt(json_web_token)
+        return jwt.decode(
+            json_web_token,
+            signing_key.key,
+            algorithms=self._server_config.id_token_signing_alg_values_supported,
+            verify=True,
+            audience=self._server_config.client_audience,
+            issuer=self._server_config.issuer,
+        )
+
+    def decode_token(self, token: dict[str, Any]) -> dict[str, Any]:
+        try:
+            # If the access_token is a JWT, we check if it is still valid
+            return self.decode_jwt(token["access_token"])
+        except jwt.DecodeError:
+            # Else, we check if the id_token is still valid
+            return self.decode_jwt(token["id_token"])
```

Comment:
Previously, this was used to notify the user that the token for "John Doe" failed to load and that they were logged in as "John Doe."  But I think this can be removed now. The assumption is that if the user provides their username and password, they are aware of whom they are attempting to log in as.